### PR TITLE
133 address book

### PR DESCRIPTION
**Pull request by rkzel**

### CODE REVIEW

File: apps/allocations/app/components/Form/Input/OptionsInput.js [hunk 1]
```diff
@@ -18,14 +18,17 @@ class OptionsInput extends React.Component {
...
   }
 
   addOption = () => {
     // TODO: Implement some rules about what an 'Option can be' duplicates, etc
     const { input, name, value } = this.props
-    if (input && !value.includes(input)) {
+    if (input && !value.includes(input) && this.props.validator(input)) {
       this.props.onChange({ target: { name, value: [...value, input] } })
       this.props.onChange({ target: { name: 'optionsInput', value: '' } })
+      this.props.error = true
```
Comment (sohkai) [inline apps/allocations/app/components/Form/Input/OptionsInput.js]: You generally don't want to do this; you probably want to have a component set its state instead (changing the prop like this doesn't cause a rerender, and the next rerender might overwrite these existing props leading to this state being lost).

File: apps/allocations/app/components/Panel/NewAllocation.js [hunk 2]
```diff
@@ -1,16 +1,9 @@
 import PropTypes from 'prop-types'
 import React from 'react'
 import styled from 'styled-components'
+import { BigNumber } from 'bignumber.js'
```
Comment (sohkai) [inline apps/allocations/app/components/Panel/NewAllocation.js]: Not a big deal ATM, but highly recommend using [BN.js](https://github.com/indutny/bn.js/) instead.

File: apps/allocations/app/components/Panel/NewAllocation.js [hunk 3]
```diff
@@ -66,23 +59,48 @@ class NewAllocation extends React.Component {
...
     // clear input here.
+
     let informational = (this.state.allocationTypeIndex === 0)
     let recurring = !informational && this.state.payoutTypeIndex != 0
     // TODO: period should be smarter: now the only option is monthly
     let period = recurring ? 86400 * 31 : 0
-
+    let optionsInput = this.state.optionsInput
+    this.setState({
+      addressError: this.state.addressError,
```
Comment (sohkai) [inline apps/allocations/app/components/Panel/NewAllocation.js]: This probably doesn't need to be reset here

File: apps/allocations/app/components/Panel/NewAllocation.js [hunk 4]
```diff
@@ -66,23 +59,48 @@ class NewAllocation extends React.Component {
...
+      allocationError: false,
+    })
+    if (!(this.isAddress(optionsInput) || optionsInput === '')) {
+      this.setState({ addressError: true })
+      this.setState()
```
Comment (sohkai) [inline apps/allocations/app/components/Panel/NewAllocation.js]: Could probably remove this extra `setState()`

File: apps/allocations/app/components/Panel/NewAllocation.js [hunk 5]
```diff
@@ -66,23 +59,48 @@ class NewAllocation extends React.Component {
...
+      return
+    }
+    if (this.isAddress(optionsInput)) {
+      this.state.options.push(this.state.optionsInput)
```
Comment (sohkai) [inline apps/allocations/app/components/Panel/NewAllocation.js]: What you probably want here is to use the [function argument form of `setState()`](https://reactjs.org/docs/state-and-lifecycle.html#state-updates-may-be-asynchronous) instead (and as a rule, **never** directly modify state).

File: apps/allocations/app/components/Panel/NewAllocation.js [hunk 6]
```diff
@@ -66,23 +59,48 @@ class NewAllocation extends React.Component {
...
-      balance: this.state.amount,
+      balance: this.state.amount * 10e17,
+    }
+    if (allocation.balance > this.props.limit && !informational) {
+      this.setState({ allocationError: true })
+      return
     }
-
     this.props.onSubmitAllocation(allocation)
+    this.state.allocationError = false
```
Comment (sohkai) [inline apps/allocations/app/components/Panel/NewAllocation.js]: See above

File: apps/allocations/app/components/Panel/NewAllocation.js (hunk 2 above)
Comment (Quazia) [reply apps/allocations/app/components/Panel/NewAllocation.js]: Why exactly do you recommand BN.js, in my experience BigNumber has a richer feature set.

File: apps/allocations/app/components/Form/Input/OptionsInput.js (hunk 1 above)
Comment (ottodevs) [reply apps/allocations/app/components/Form/Input/OptionsInput.js]: OptionsInput was refactored to SFC and props were reviewed

File: apps/allocations/app/components/Panel/NewAllocation.js (hunk 3 above)
Comment (ottodevs) [reply apps/allocations/app/components/Panel/NewAllocation.js]: Error system at NewAllocation panel was completely refactored

File: apps/allocations/app/components/Panel/NewAllocation.js (hunk 4 above)
Comment (ottodevs) [reply apps/allocations/app/components/Panel/NewAllocation.js]: Done

File: apps/allocations/app/components/Panel/NewAllocation.js (hunk 5 above)
Comment (ottodevs) [reply apps/allocations/app/components/Panel/NewAllocation.js]: Fixed

File: apps/allocations/app/components/Panel/NewAllocation.js (hunk 6 above)
Comment (ottodevs) [reply apps/allocations/app/components/Panel/NewAllocation.js]: 👍 fixed

**Comment by ottodevs** [comment]
All review comments were addresses, it remains pending the decision to switch to BN.js

File: apps/allocations/app/components/Panel/NewAllocation.js (hunk 2 above)
Comment (sohkai) [reply apps/allocations/app/components/Panel/NewAllocation.js]: That's true, there are specific cases where `BigNumber.js` is much easier to work with, but `BN.js` is faster and most implementations are moving to it.

**Pull request merged**